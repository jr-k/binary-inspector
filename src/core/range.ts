import { assert } from "chai";

/**
 * This represents a byte range in an ArrayBuffer. This can be used to read out
 * data as well as for highlighting regions in the BinaryView.
 */
export class ByteRange {
  buffer: ArrayBuffer;
  byteStart: number;
  byteLength: number;

  constructor(buffer: ArrayBuffer, byteStart: number, byteLength: number) {
    this.buffer = buffer;
    this.byteStart = byteStart;
    this.byteLength = byteLength;
  }

  toDataView() {
    return new DataView(this.buffer, this.byteStart, this.byteLength);
  }

  toUint8Array() {
    return new Uint8Array(this.buffer, this.byteStart, this.byteLength);
  }

  toHex(): string {
    return Array.from(this.toUint8Array())
      .map(b => b.toString(16).padStart(2, "0"))
      .join("");
  }

  bytes(byteStart: number, byteLength?: number) {
    if (byteLength == null) {
      byteLength = this.byteLength - byteStart;
    }

    if (byteStart < 0)
      throw new Error(`byteStart is ${byteStart}, but should be non-negative.`);
    if (byteStart + byteLength > this.byteLength)
      throw new Error(`New subrange does not fit within current subrange.`);

    return new ByteRange(this.buffer, this.byteStart + byteStart, byteLength);
  }

  bits(bitStart: number, bits: number) {
    return new BitRange(this.buffer, this.byteStart * 8 + bitStart, bits);
  }

  readBits(): Array<boolean> {
    let bits: Array<boolean> = [];
    let bytes = this.toUint8Array();
    while (bits.length < this.byteLength * 8) {
      let byte = bytes[Math.floor(bits.length / 8)];
      let pos = bits.length % 8;

      bits.push(((byte << pos) & 128) === 128);
    }
    return bits;
  }

  chunks(size: number): Array<ByteRange> {
    let chunks: Array<ByteRange> = [];
    let cursor = 0;
    while (cursor < this.byteLength) {
      chunks.push(this.bytes(cursor, Math.min(size, this.byteLength - cursor)));
      cursor += size;
    }
    return chunks;
  }
}

/**
 * This represents a range of bits in an ArrayBuffer. Use for reading bitfields.
 */
export class BitRange {
  buffer: ArrayBuffer;
  bitStart: number;
  bits: number;

  constructor(buffer: ArrayBuffer, bitStart: number, bits: number) {
    this.buffer = buffer;
    this.bitStart = bitStart;
    this.bits = bits;
  }

  enclosingByteRange(): ByteRange {
    return new ByteRange(
      this.buffer,
      Math.floor(this.bitStart / 8),
      Math.floor((this.bitStart + this.bits - 1) / 8) + 1
    );
  }

  readBits(): Array<boolean> {
    let bits = this.enclosingByteRange().readBits();
    return bits.slice(this.bitStart, this.bitStart + this.bits);
  }

  readUIntBE(): number {
    let number = 0;
    let bits = this.readBits();
    for (let i = 0; i < bits.length; ++i) {
      if (bits[bits.length - 1 - i]) number += Math.pow(2, i);
    }
    return number;
  }

  readBool(): boolean {
    let bits = this.readBits();
    assert.equal(bits.length, 1);
    return bits[0];
  }
}
