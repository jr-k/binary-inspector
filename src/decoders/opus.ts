import { ByteRange, BitRange } from "../core/range";
import { Tree } from "../core/tree";
import { assert } from "chai";

enum Mode {
  SILK_ONLY = "SILK-only",
  HYBRID = "Hybrid",
  CELT_ONLY = "CELT-only"
}

enum Bandwidth {
  NB = "NB",
  MB = "MB",
  WB = "WB",
  SWB = "SWB",
  FB = "FB"
}

function getSampleRate(bandwidth: Bandwidth): number {
  switch (bandwidth) {
    case Bandwidth.NB:
      return 8000;
    case Bandwidth.MB:
      return 12000;
    case Bandwidth.WB:
      return 16000;
    case Bandwidth.SWB:
      return 24000;
    case Bandwidth.FB:
      return 48000;
  }
}

function getBandwidth(bandwidth: Bandwidth): number {
  switch (bandwidth) {
    case Bandwidth.NB:
      return 4000;
    case Bandwidth.MB:
      return 6000;
    case Bandwidth.WB:
      return 8000;
    case Bandwidth.SWB:
      return 12000;
    case Bandwidth.FB:
      return 20000;
  }
}

const SILK_ONLY_FRAME_SIZES = [10, 20, 40, 60];
const HYBRID_FRAME_SIZES = [10, 20];
const CELT_ONLY_FRAME_SIZES = [2.5, 5, 10, 20];

type Config = [Mode, Bandwidth, number];

function parseConfig(config): Config {
  if (config <= 3) {
    return [Mode.SILK_ONLY, Bandwidth.NB, SILK_ONLY_FRAME_SIZES[config]];
  } else if (config <= 7) {
    return [Mode.SILK_ONLY, Bandwidth.MB, SILK_ONLY_FRAME_SIZES[config - 4]];
  } else if (config <= 11) {
    return [Mode.SILK_ONLY, Bandwidth.WB, SILK_ONLY_FRAME_SIZES[config - 8]];
  } else if (config <= 13) {
    return [Mode.HYBRID, Bandwidth.SWB, HYBRID_FRAME_SIZES[config - 12]];
  } else if (config <= 15) {
    return [Mode.HYBRID, Bandwidth.FB, HYBRID_FRAME_SIZES[config - 14]];
  } else if (config <= 19) {
    return [Mode.CELT_ONLY, Bandwidth.NB, CELT_ONLY_FRAME_SIZES[config - 16]];
  } else if (config <= 23) {
    return [Mode.CELT_ONLY, Bandwidth.WB, CELT_ONLY_FRAME_SIZES[config - 20]];
  } else if (config <= 27) {
    return [Mode.CELT_ONLY, Bandwidth.SWB, CELT_ONLY_FRAME_SIZES[config - 24]];
  } else if (config <= 31) {
    return [Mode.CELT_ONLY, Bandwidth.FB, CELT_ONLY_FRAME_SIZES[config - 28]];
  } else {
    throw new Error(`Unknown config ${config}.`);
  }
}

function inspectConfig(config: BitRange): [Tree, Config] {
  let [mode, bw, fs] = parseConfig(config.readUIntBE());
  return [
    new Tree(
      `Config: ${config.readUIntBE()} (${mode}, ${bw}, ${fs}ms)`,
      config,
      [
        new Tree(`Mode: ${mode}`, config),
        new Tree(
          `Bandwidth: ${bw} (${getBandwidth(
            bw
          )}Hz), Sample Rate: ${getSampleRate(bw)}Hz`,
          config
        ),
        new Tree(`Frame Size: ${fs}ms`, config)
      ]
    ),
    [mode, bw, fs]
  ];
}

function getSILKFrameSize(opusFrameSize: number): number {
  switch (opusFrameSize) {
    case 10:
      return 10;
    case 20:
      return 20;
    case 40:
      return 20;
    case 60:
      return 20;
    default:
      throw new Error(`Unexpected Opus frame size ${opusFrameSize}.`);
  }
}

function inspectSILKFrame(
  range: ByteRange,
  opusFrameSize: number,
  stereo: boolean
): Tree {
  let silkFrameSize = getSILKFrameSize(opusFrameSize);
  let silkFramesCount = opusFrameSize / silkFrameSize;
  let channelCount = stereo ? 2 : 1;

  let channelNames = stereo ? ["Mid", "Side"] : ["Mono"];

  let headerTree: Array<Tree> = [];
  let ptr = range.bits(0);
  for (let i = 0; i < channelCount; ++i) {
    for (let j = 0; j < silkFramesCount; ++j) {
      // Read one bit.
      let vadFlag = ptr.bits(0, 1);
      ptr = ptr.bits(1);

      headerTree.push(
        new Tree(
          `${channelNames[i]}, Frame ${j + 1}, VAD Flag: ${vadFlag.readBool()}`,
          vadFlag
        )
      );
    }

    // Read one bit.
    let lbrrFlag = ptr.bits(0, 1);
    ptr = ptr.bits(1);

    headerTree.push(
      new Tree(
        `${channelNames[i]}, LBRR Flag: ${lbrrFlag.readBool()}`,
        lbrrFlag
      )
    );
  }

  return new Tree(
    `SILK Frame (${silkFramesCount} SILK Frames)`,
    range,
    headerTree
  );
}

function inspectFrame(
  range: ByteRange,
  mode: Mode,
  frameSize: number,
  stereo: boolean
): Tree {
  if (mode == Mode.SILK_ONLY) {
    return inspectSILKFrame(range, frameSize, stereo);
  }
  if (mode == Mode.HYBRID) {
    return new Tree(`Hybrid Frame`, range, [
      inspectSILKFrame(range, frameSize, stereo)
    ]);
  } else {
    return new Tree(`Compressed Frame (${range.size()} bytes)`, range, []);
  }
}

function readFrameLength(range: ByteRange): [number, ByteRange] {
  let firstByte = range.bytes(0, 1).readUIntBE();
  if (firstByte <= 251) {
    return [firstByte, range.bytes(0, 1)];
  } else {
    let secondByte = range.bytes(1, 1).readUIntBE();
    return [secondByte * 4 + firstByte, range.bytes(0, 2)];
  }
}

function inspectFrames(
  c: number,
  range: ByteRange,
  mode: Mode,
  frameSize: number,
  stereo: boolean
): Tree {
  let frames: Array<Tree> = [];

  try {
    if (c == 0) {
      frames = [inspectFrame(range, mode, frameSize, stereo)];
    } else if (c == 1) {
      assert(
        range.size() % 2 == 0,
        "In packing mode 2, the amount of bytes available for frames must be even."
      );
      frames = [
        inspectFrame(
          range.bytes(0, range.byteLength / 2),
          mode,
          frameSize,
          stereo
        ),
        inspectFrame(range.bytes(range.byteLength / 2), mode, frameSize, stereo)
      ];
    } else if (c == 2) {
      assert.fail("Packing mode 2 unimplemented.");
    } else if (c == 3) {
      let info = range.bytes(0, 1);
      let v = info.bits(0, 1);
      let p = info.bits(1, 1);
      let M = info.bits(2, 6);

      frames.push(new Tree(`Mode: ${v.readBool() ? "VBR" : "CBR"}`, v));
      frames.push(new Tree(`Padding: ${p.readBool()}`, p));
      frames.push(new Tree(`Frame Count: ${M.readUIntBE()}`, M));

      if (p.readBool()) assert.fail("Frame padding not implemented.");
      if (v.readBool()) {
        let frameLengths: Array<[number, ByteRange]> = [];
        let pointer = range.bytes(1);
        for (let i = 0; i < M.readUIntBE() - 1; ++i) {
          let [frameLength, frameLengthRange] = readFrameLength(pointer);
          frameLengths.push([frameLength, frameLengthRange]);
          pointer = pointer.bytes(frameLengthRange.size());
        }

        frames.push(
          new Tree(
            `Frame Lengths: ${frameLengths.length}`,
            range.bytes(1, pointer.byteStart - frameLengths[0][1].byteStart),
            frameLengths.map(
              f => new Tree(`${f[1].size()}-byte Frame Length: ${f[0]}`, f[1])
            )
          )
        );

        for (let [frameLength, _] of frameLengths) {
          let frameRange = pointer.bytes(0, frameLength);
          frames.push(inspectFrame(frameRange, mode, frameSize, stereo));

          pointer = pointer.bytes(frameLength);
        }

        let frameRange = pointer;
        frames.push(inspectFrame(frameRange, mode, frameSize, stereo));
      } else {
        assert.fail("CBR mode not implemented.");
      }
    }

    return new Tree(`Compressed Frames`, range, frames);
  } catch (e) {
    return new Tree(`Malformed Frames: ${e.message}`, range, frames, e);
  }
}

export function inspect(range: ByteRange): Tree {
  assert.isAtLeast(range.size(), 1, "Opus packet must be at least one byte.");

  let toc = range.bytes(0, 1);
  let config = toc.bits(0, 5);
  let s = toc.bits(5, 1);
  let c = toc.bits(6, 2);

  let packingCodes = [
    "1 Frame",
    "2 Frame, Equal Size",
    "2 Frames, Different Size",
    "Arbitrary Frames"
  ];

  let [configTree, [mode, bandwidth, frameSize]] = inspectConfig(config);

  return new Tree(`Opus Packet`, range, [
    configTree,
    new Tree(`Channels: ${s.readBool() ? "Stereo" : "Mono"}`, s),
    new Tree(
      `Frame Packing Mode: ${c.readUIntBE()} (${packingCodes[c.readUIntBE()]})`,
      c
    ),
    inspectFrames(c.readUIntBE(), range.bytes(1), mode, frameSize, s.readBool())
  ]);
}
