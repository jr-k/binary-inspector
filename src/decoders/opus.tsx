import { ByteRange, BitRange } from "../core/range";
import { Tree } from "../core/tree";
import { assert } from "chai";

enum Mode {
  SILK_ONLY = "SILK-only",
  HYBRID = "Hybrid",
  CELT_ONLY = "CELT-only"
}

enum Bandwidth {
  NB = "NB",
  MB = "MB",
  WB = "WB",
  SWB = "SWB",
  FB = "FB"
}

function getSampleRate(bandwidth: Bandwidth): number {
  switch (bandwidth) {
    case Bandwidth.NB:
      return 8000;
    case Bandwidth.MB:
      return 12000;
    case Bandwidth.WB:
      return 16000;
    case Bandwidth.SWB:
      return 24000;
    case Bandwidth.FB:
      return 48000;
  }
}

function getBandwidth(bandwidth: Bandwidth): number {
  switch (bandwidth) {
    case Bandwidth.NB:
      return 4000;
    case Bandwidth.MB:
      return 6000;
    case Bandwidth.WB:
      return 8000;
    case Bandwidth.SWB:
      return 12000;
    case Bandwidth.FB:
      return 20000;
  }
}

const SILK_ONLY_FRAME_SIZES = [10, 20, 40, 60];
const HYBRID_FRAME_SIZES = [10, 20];
const CELT_ONLY_FRAME_SIZES = [2.5, 5, 10, 20];

function parseConfig(config): [Mode, Bandwidth, number] {
  if (config <= 3) {
    return [Mode.SILK_ONLY, Bandwidth.NB, SILK_ONLY_FRAME_SIZES[config]];
  } else if (config <= 7) {
    return [Mode.SILK_ONLY, Bandwidth.MB, SILK_ONLY_FRAME_SIZES[config - 4]];
  } else if (config <= 11) {
    return [Mode.SILK_ONLY, Bandwidth.WB, SILK_ONLY_FRAME_SIZES[config - 8]];
  } else if (config <= 13) {
    return [Mode.HYBRID, Bandwidth.SWB, HYBRID_FRAME_SIZES[config - 12]];
  } else if (config <= 15) {
    return [Mode.HYBRID, Bandwidth.FB, HYBRID_FRAME_SIZES[config - 14]];
  } else if (config <= 19) {
    return [Mode.CELT_ONLY, Bandwidth.NB, CELT_ONLY_FRAME_SIZES[config - 16]];
  } else if (config <= 23) {
    return [Mode.CELT_ONLY, Bandwidth.WB, CELT_ONLY_FRAME_SIZES[config - 20]];
  } else if (config <= 27) {
    return [Mode.CELT_ONLY, Bandwidth.SWB, CELT_ONLY_FRAME_SIZES[config - 24]];
  } else if (config <= 31) {
    return [Mode.CELT_ONLY, Bandwidth.FB, CELT_ONLY_FRAME_SIZES[config - 28]];
  } else {
    throw new Error(`Unknown config ${config}.`);
  }
}

function inspectConfig(config: BitRange): Tree {
  let [mode, bw, fs] = parseConfig(config.readUIntBE());
  return new Tree(
    `Config: ${config.readUIntBE()} (${mode}, ${bw}, ${fs}ms)`,
    config,
    [
      new Tree(`Mode: ${mode}`, config),
      new Tree(
        `Bandwidth: ${bw} (${getBandwidth(bw)}Hz), Sample Rate: ${getSampleRate(
          bw
        )}Hz`,
        config
      ),
      new Tree(`Frame Size: ${fs}ms`, config)
    ]
  );
}

function inspectFrame(range: ByteRange): Tree {
  return new Tree(`Compressed Frame`, range, []);
}

function inspectFrames(c: number, range: ByteRange): Tree {
  let frames: Array<Tree> = [];
  if (c == 0) {
    frames = [inspectFrame(range)];
  } else if (c == 1) {
    assert(range.byteLength % 2 == 0);
    frames = [
      inspectFrame(range.bytes(0, range.byteLength / 2)),
      inspectFrame(range.bytes(range.byteLength / 2))
    ];
  }

  return new Tree(`Compressed Frames`, range, frames);
}

export function inspect(range: ByteRange): Tree {
  let toc = range.bytes(0, 1);
  let config = toc.bits(0, 5);
  let s = toc.bits(5, 1);
  let c = toc.bits(6, 2);

  let packingCodes = [
    "1 Frame",
    "2 Frame, Equal Size",
    "2 Frames, Different Size",
    "Arbitrary Frames"
  ];

  return new Tree(`Opus Packet`, range, [
    inspectConfig(config),
    new Tree(`Channels: ${s.readBool() ? "Stereo" : "Mono"}`, s),
    new Tree(
      `Frame Packing Mode: ${c.readUIntBE()} (${packingCodes[c.readUIntBE()]})`,
      c
    ),
    inspectFrames(c.readUIntBE(), range.bytes(1))
  ]);
}
