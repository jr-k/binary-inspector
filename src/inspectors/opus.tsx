import * as React from "react";
import * as ReactDOM from "react-dom";

import { SimpleInspector } from "../ui/simpleinspector";
import { Container } from "react-bootstrap";
import { inspect } from "../decoders/opus";

ReactDOM.render(
  <SimpleInspector inspect={inspect} />,
  document.getElementById("root")
);
