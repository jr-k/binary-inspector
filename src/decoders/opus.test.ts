import { inspect } from "./opus";
import { base64ToArrayBuffer } from "../core/utils";
import { ByteRange } from "../core/range";

const PACKET_A = base64ToArrayBuffer(
  "S4gvMC0sLyMlgPH9SANsMHykzdeG0IgL2AOMLDZENcJeFbzQNtjpEFr/n+hdoNR2GW/RMhKIa0mDy3V3mkYcHiblcjazYthzd+7bjEX0y/NR8T66tFPxsNqQbafdmTrOWjzwJklaWSCEXAAYoTvZEvE6v+r9Z42VvDd+VkPJafZKZOZyNaiH3I7DSP/eH6GEb0ztpFCKoFrSTtVNf2L/BiyXSjWD9GLrDW2nqTO/H9jJLDq+T1M2O8Ytq7eDorjyX4fya7sOtcmXA32VrYBq98uzZE0RqHo6JGtLKFcYIwOiVgSA3Hr2/ZeYYLQuWBzgKN3WCpwJU9GXx/OAicGwBscuhRN0iY4ybT7fQtsl4F9S5jYkCDU3eC2P0NZgt+7lRax4/gAmQv9PXb2HCSlojEj+3Rl3c7HwHlQtiK2uMP9bwBEUV42ehGybc74i3YrE1c4L5DngOxuhSiSA"
);

const PACKET_B = base64ToArrayBuffer(
  "eC1O601F53FIhPJqoFlaD707r0cmq8YYzkSV7zWNXFc3YgYUc3Qq7IQkAzTIfS3nfPZsFPQVbh7o5NrLeg5DEEuCVh8prlP6"
);

test("inspect", () => {
  let packet_a = new ByteRange(PACKET_A, 0, PACKET_A.byteLength);
  expect(inspect(packet_a)).toMatchSnapshot();

  let packet_b = new ByteRange(PACKET_B, 0, PACKET_B.byteLength);
  expect(inspect(packet_b)).toMatchSnapshot();
});
