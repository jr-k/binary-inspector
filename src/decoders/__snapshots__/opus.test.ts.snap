// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inspect 1`] = `
Tree {
  "children": Array [
    Tree {
      "children": Array [
        Tree {
          "children": Array [],
          "error": undefined,
          "label": "Mode: SILK-only",
          "parent": [Circular],
          "range": BitRange {
            "bitLength": 5,
            "bitStart": 0,
            "buffer": ArrayBuffer [],
          },
        },
        Tree {
          "children": Array [],
          "error": undefined,
          "label": "Bandwidth: WB (8000Hz), Sample Rate: 16000Hz",
          "parent": [Circular],
          "range": BitRange {
            "bitLength": 5,
            "bitStart": 0,
            "buffer": ArrayBuffer [],
          },
        },
        Tree {
          "children": Array [],
          "error": undefined,
          "label": "Frame Size: 20ms",
          "parent": [Circular],
          "range": BitRange {
            "bitLength": 5,
            "bitStart": 0,
            "buffer": ArrayBuffer [],
          },
        },
      ],
      "error": undefined,
      "label": "Config: 9 (SILK-only, WB, 20ms)",
      "parent": [Circular],
      "range": BitRange {
        "bitLength": 5,
        "bitStart": 0,
        "buffer": ArrayBuffer [],
      },
    },
    Tree {
      "children": Array [],
      "error": undefined,
      "label": "Channels: Mono",
      "parent": [Circular],
      "range": BitRange {
        "bitLength": 1,
        "bitStart": 5,
        "buffer": ArrayBuffer [],
      },
    },
    Tree {
      "children": Array [],
      "error": undefined,
      "label": "Frame Packing Mode: 3 (Arbitrary Frames)",
      "parent": [Circular],
      "range": BitRange {
        "bitLength": 2,
        "bitStart": 6,
        "buffer": ArrayBuffer [],
      },
    },
    Tree {
      "children": Array [
        Tree {
          "children": Array [],
          "error": undefined,
          "label": "Mode: VBR",
          "parent": [Circular],
          "range": BitRange {
            "bitLength": 1,
            "bitStart": 8,
            "buffer": ArrayBuffer [],
          },
        },
        Tree {
          "children": Array [],
          "error": undefined,
          "label": "Padding: false",
          "parent": [Circular],
          "range": BitRange {
            "bitLength": 1,
            "bitStart": 9,
            "buffer": ArrayBuffer [],
          },
        },
        Tree {
          "children": Array [],
          "error": undefined,
          "label": "Frame Count: 8",
          "parent": [Circular],
          "range": BitRange {
            "bitLength": 6,
            "bitStart": 10,
            "buffer": ArrayBuffer [],
          },
        },
        Tree {
          "children": Array [
            Tree {
              "children": Array [],
              "error": undefined,
              "label": "1-byte Frame Length: 47",
              "parent": [Circular],
              "range": ByteRange {
                "buffer": ArrayBuffer [],
                "byteLength": 1,
                "byteStart": 2,
              },
            },
            Tree {
              "children": Array [],
              "error": undefined,
              "label": "1-byte Frame Length: 48",
              "parent": [Circular],
              "range": ByteRange {
                "buffer": ArrayBuffer [],
                "byteLength": 1,
                "byteStart": 3,
              },
            },
            Tree {
              "children": Array [],
              "error": undefined,
              "label": "1-byte Frame Length: 45",
              "parent": [Circular],
              "range": ByteRange {
                "buffer": ArrayBuffer [],
                "byteLength": 1,
                "byteStart": 4,
              },
            },
            Tree {
              "children": Array [],
              "error": undefined,
              "label": "1-byte Frame Length: 44",
              "parent": [Circular],
              "range": ByteRange {
                "buffer": ArrayBuffer [],
                "byteLength": 1,
                "byteStart": 5,
              },
            },
            Tree {
              "children": Array [],
              "error": undefined,
              "label": "1-byte Frame Length: 47",
              "parent": [Circular],
              "range": ByteRange {
                "buffer": ArrayBuffer [],
                "byteLength": 1,
                "byteStart": 6,
              },
            },
            Tree {
              "children": Array [],
              "error": undefined,
              "label": "1-byte Frame Length: 35",
              "parent": [Circular],
              "range": ByteRange {
                "buffer": ArrayBuffer [],
                "byteLength": 1,
                "byteStart": 7,
              },
            },
            Tree {
              "children": Array [],
              "error": undefined,
              "label": "1-byte Frame Length: 37",
              "parent": [Circular],
              "range": ByteRange {
                "buffer": ArrayBuffer [],
                "byteLength": 1,
                "byteStart": 8,
              },
            },
          ],
          "error": undefined,
          "label": "Frame Lengths: 7",
          "parent": [Circular],
          "range": ByteRange {
            "buffer": ArrayBuffer [],
            "byteLength": 7,
            "byteStart": 2,
          },
        },
        Tree {
          "children": Array [
            Tree {
              "children": Array [],
              "error": undefined,
              "label": "Mono, Frame 1, VAD Flag: true",
              "parent": [Circular],
              "range": BitRange {
                "bitLength": 1,
                "bitStart": 72,
                "buffer": ArrayBuffer [],
              },
            },
            Tree {
              "children": Array [],
              "error": undefined,
              "label": "Mono, LBRR Flag: false",
              "parent": [Circular],
              "range": BitRange {
                "bitLength": 1,
                "bitStart": 73,
                "buffer": ArrayBuffer [],
              },
            },
          ],
          "error": undefined,
          "label": "SILK Frame (1 SILK Frames)",
          "parent": [Circular],
          "range": ByteRange {
            "buffer": ArrayBuffer [],
            "byteLength": 47,
            "byteStart": 9,
          },
        },
        Tree {
          "children": Array [
            Tree {
              "children": Array [],
              "error": undefined,
              "label": "Mono, Frame 1, VAD Flag: true",
              "parent": [Circular],
              "range": BitRange {
                "bitLength": 1,
                "bitStart": 448,
                "buffer": ArrayBuffer [],
              },
            },
            Tree {
              "children": Array [],
              "error": undefined,
              "label": "Mono, LBRR Flag: false",
              "parent": [Circular],
              "range": BitRange {
                "bitLength": 1,
                "bitStart": 449,
                "buffer": ArrayBuffer [],
              },
            },
          ],
          "error": undefined,
          "label": "SILK Frame (1 SILK Frames)",
          "parent": [Circular],
          "range": ByteRange {
            "buffer": ArrayBuffer [],
            "byteLength": 48,
            "byteStart": 56,
          },
        },
        Tree {
          "children": Array [
            Tree {
              "children": Array [],
              "error": undefined,
              "label": "Mono, Frame 1, VAD Flag: true",
              "parent": [Circular],
              "range": BitRange {
                "bitLength": 1,
                "bitStart": 832,
                "buffer": ArrayBuffer [],
              },
            },
            Tree {
              "children": Array [],
              "error": undefined,
              "label": "Mono, LBRR Flag: false",
              "parent": [Circular],
              "range": BitRange {
                "bitLength": 1,
                "bitStart": 833,
                "buffer": ArrayBuffer [],
              },
            },
          ],
          "error": undefined,
          "label": "SILK Frame (1 SILK Frames)",
          "parent": [Circular],
          "range": ByteRange {
            "buffer": ArrayBuffer [],
            "byteLength": 45,
            "byteStart": 104,
          },
        },
        Tree {
          "children": Array [
            Tree {
              "children": Array [],
              "error": undefined,
              "label": "Mono, Frame 1, VAD Flag: true",
              "parent": [Circular],
              "range": BitRange {
                "bitLength": 1,
                "bitStart": 1192,
                "buffer": ArrayBuffer [],
              },
            },
            Tree {
              "children": Array [],
              "error": undefined,
              "label": "Mono, LBRR Flag: false",
              "parent": [Circular],
              "range": BitRange {
                "bitLength": 1,
                "bitStart": 1193,
                "buffer": ArrayBuffer [],
              },
            },
          ],
          "error": undefined,
          "label": "SILK Frame (1 SILK Frames)",
          "parent": [Circular],
          "range": ByteRange {
            "buffer": ArrayBuffer [],
            "byteLength": 44,
            "byteStart": 149,
          },
        },
        Tree {
          "children": Array [
            Tree {
              "children": Array [],
              "error": undefined,
              "label": "Mono, Frame 1, VAD Flag: true",
              "parent": [Circular],
              "range": BitRange {
                "bitLength": 1,
                "bitStart": 1544,
                "buffer": ArrayBuffer [],
              },
            },
            Tree {
              "children": Array [],
              "error": undefined,
              "label": "Mono, LBRR Flag: false",
              "parent": [Circular],
              "range": BitRange {
                "bitLength": 1,
                "bitStart": 1545,
                "buffer": ArrayBuffer [],
              },
            },
          ],
          "error": undefined,
          "label": "SILK Frame (1 SILK Frames)",
          "parent": [Circular],
          "range": ByteRange {
            "buffer": ArrayBuffer [],
            "byteLength": 47,
            "byteStart": 193,
          },
        },
        Tree {
          "children": Array [
            Tree {
              "children": Array [],
              "error": undefined,
              "label": "Mono, Frame 1, VAD Flag: false",
              "parent": [Circular],
              "range": BitRange {
                "bitLength": 1,
                "bitStart": 1920,
                "buffer": ArrayBuffer [],
              },
            },
            Tree {
              "children": Array [],
              "error": undefined,
              "label": "Mono, LBRR Flag: false",
              "parent": [Circular],
              "range": BitRange {
                "bitLength": 1,
                "bitStart": 1921,
                "buffer": ArrayBuffer [],
              },
            },
          ],
          "error": undefined,
          "label": "SILK Frame (1 SILK Frames)",
          "parent": [Circular],
          "range": ByteRange {
            "buffer": ArrayBuffer [],
            "byteLength": 35,
            "byteStart": 240,
          },
        },
        Tree {
          "children": Array [
            Tree {
              "children": Array [],
              "error": undefined,
              "label": "Mono, Frame 1, VAD Flag: false",
              "parent": [Circular],
              "range": BitRange {
                "bitLength": 1,
                "bitStart": 2200,
                "buffer": ArrayBuffer [],
              },
            },
            Tree {
              "children": Array [],
              "error": undefined,
              "label": "Mono, LBRR Flag: false",
              "parent": [Circular],
              "range": BitRange {
                "bitLength": 1,
                "bitStart": 2201,
                "buffer": ArrayBuffer [],
              },
            },
          ],
          "error": undefined,
          "label": "SILK Frame (1 SILK Frames)",
          "parent": [Circular],
          "range": ByteRange {
            "buffer": ArrayBuffer [],
            "byteLength": 37,
            "byteStart": 275,
          },
        },
        Tree {
          "children": Array [
            Tree {
              "children": Array [],
              "error": undefined,
              "label": "Mono, Frame 1, VAD Flag: false",
              "parent": [Circular],
              "range": BitRange {
                "bitLength": 1,
                "bitStart": 2496,
                "buffer": ArrayBuffer [],
              },
            },
            Tree {
              "children": Array [],
              "error": undefined,
              "label": "Mono, LBRR Flag: false",
              "parent": [Circular],
              "range": BitRange {
                "bitLength": 1,
                "bitStart": 2497,
                "buffer": ArrayBuffer [],
              },
            },
          ],
          "error": undefined,
          "label": "SILK Frame (1 SILK Frames)",
          "parent": [Circular],
          "range": ByteRange {
            "buffer": ArrayBuffer [],
            "byteLength": 36,
            "byteStart": 312,
          },
        },
      ],
      "error": undefined,
      "label": "Compressed Frames",
      "parent": [Circular],
      "range": ByteRange {
        "buffer": ArrayBuffer [],
        "byteLength": 347,
        "byteStart": 1,
      },
    },
  ],
  "error": undefined,
  "label": "Opus Packet",
  "range": ByteRange {
    "buffer": ArrayBuffer [],
    "byteLength": 348,
    "byteStart": 0,
  },
}
`;

exports[`inspect 2`] = `
Tree {
  "children": Array [
    Tree {
      "children": Array [
        Tree {
          "children": Array [],
          "error": undefined,
          "label": "Mode: Hybrid",
          "parent": [Circular],
          "range": BitRange {
            "bitLength": 5,
            "bitStart": 0,
            "buffer": ArrayBuffer [],
          },
        },
        Tree {
          "children": Array [],
          "error": undefined,
          "label": "Bandwidth: FB (20000Hz), Sample Rate: 48000Hz",
          "parent": [Circular],
          "range": BitRange {
            "bitLength": 5,
            "bitStart": 0,
            "buffer": ArrayBuffer [],
          },
        },
        Tree {
          "children": Array [],
          "error": undefined,
          "label": "Frame Size: 20ms",
          "parent": [Circular],
          "range": BitRange {
            "bitLength": 5,
            "bitStart": 0,
            "buffer": ArrayBuffer [],
          },
        },
      ],
      "error": undefined,
      "label": "Config: 15 (Hybrid, FB, 20ms)",
      "parent": [Circular],
      "range": BitRange {
        "bitLength": 5,
        "bitStart": 0,
        "buffer": ArrayBuffer [],
      },
    },
    Tree {
      "children": Array [],
      "error": undefined,
      "label": "Channels: Mono",
      "parent": [Circular],
      "range": BitRange {
        "bitLength": 1,
        "bitStart": 5,
        "buffer": ArrayBuffer [],
      },
    },
    Tree {
      "children": Array [],
      "error": undefined,
      "label": "Frame Packing Mode: 0 (1 Frame)",
      "parent": [Circular],
      "range": BitRange {
        "bitLength": 2,
        "bitStart": 6,
        "buffer": ArrayBuffer [],
      },
    },
    Tree {
      "children": Array [
        Tree {
          "children": Array [
            Tree {
              "children": Array [
                Tree {
                  "children": Array [],
                  "error": undefined,
                  "label": "Mono, Frame 1, VAD Flag: false",
                  "parent": [Circular],
                  "range": BitRange {
                    "bitLength": 1,
                    "bitStart": 8,
                    "buffer": ArrayBuffer [],
                  },
                },
                Tree {
                  "children": Array [],
                  "error": undefined,
                  "label": "Mono, LBRR Flag: false",
                  "parent": [Circular],
                  "range": BitRange {
                    "bitLength": 1,
                    "bitStart": 9,
                    "buffer": ArrayBuffer [],
                  },
                },
              ],
              "error": undefined,
              "label": "SILK Frame (1 SILK Frames)",
              "parent": [Circular],
              "range": ByteRange {
                "buffer": ArrayBuffer [],
                "byteLength": 71,
                "byteStart": 1,
              },
            },
          ],
          "error": undefined,
          "label": "Hybrid Frame",
          "parent": [Circular],
          "range": ByteRange {
            "buffer": ArrayBuffer [],
            "byteLength": 71,
            "byteStart": 1,
          },
        },
      ],
      "error": undefined,
      "label": "Compressed Frames",
      "parent": [Circular],
      "range": ByteRange {
        "buffer": ArrayBuffer [],
        "byteLength": 71,
        "byteStart": 1,
      },
    },
  ],
  "error": undefined,
  "label": "Opus Packet",
  "range": ByteRange {
    "buffer": ArrayBuffer [],
    "byteLength": 72,
    "byteStart": 0,
  },
}
`;
